cpp/src/arrow/compute/kernels/random.cc

```diff
@@ -0,0 +1,63 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <memory>
+#include <random>
+
+#include "arrow/compute/kernel.h"
+#include "arrow/compute/kernels/common.h"
+#include "arrow/compute/registry.h"
+#include "arrow/util/pcg_random.h"
+
+namespace arrow {
+namespace compute {
+namespace internal {
+namespace {
+
+// Generates a random floating point number in range [0, 1).
+template <class Rng>
+double generate_uniform(Rng& rng) {
+  double range = static_cast<double>(Rng::max() - Rng::min()) + 1;
+  double value = static_cast<double>(rng() - Rng::min());
+  return value / range;
```

Comment:
Due to floating point round-off errors, looks we cannot guarantee the result will be strictly less than 1. [1]
Numpy's strange equation `return (rnd >> 11) * (1.0 / 9007199254740992.0)` is safer.

[1] https://godbolt.org/z/bYneE8jhr

Reply:
Ok, I've changed it to use that formula. Looks like we inherited this issue from C++ STL: https://cplusplus.github.io/LWG/lwg-active.html#2524.